src/NHibernate/Proxy/ProxyBuilderHelper.cs

```diff
@@ -253,5 +287,72 @@ private static string[] GenerateTypeNames(int count)
 
 			return result;
 		}
+
+		/// <summary>
+		/// Method equality for the proxy building purpose: we want to equate an interface method to a base type
+		/// method which implements it. This implies the base type method has the same signature and there is no
+		/// explicit implementation of the interface method in the base type.
+		/// </summary>
+		private class MethodInfoComparer : IEqualityComparer<MethodInfo>
```

Comment:
This comparer is just an optimization to generate only one method if the interface method is implemented implicitly. Right?

Reply:
Initially I wanted to handle it fully in the method using that comparer, but it was too complex.

Yes it is for generating only one method if the interface is implemented implicitly. No it is not an optimization from my viewpoint.

Not doing so, while implementing the base method and the interface method both publicly, is #2085 bug. Granted, the other changes in the method generation is enough to fix it alone.

But then, not doing so would trigger again #2052. This bug was re-qualified as just a "possible breaking change" bound to the switch to the static proxy, but if we can avoid that breaking change, that would be better.